File: .github/workflows/utils/screenshots.py
from playwright.sync_api import sync_playwright
import sys
from time import sleep

base_url = sys.argv[1]
endpoints = {
    "Login": "login",
    "RouteList": "1d3dc3e03047b0c7",
    "RouteActivity": "1d3dc3e03047b0c7/000000dd--455f14369d",
}

def take_screenshots(device_type, browser):
    page = browser.new_page()
    page.goto(f"{base_url}")
    page.get_by_role("button", name="Try the demo").click()
    for endpoint in endpoints:
        page.goto(f"{base_url}/{endpoints[endpoint]}")
        sleep(2)
        page.screenshot(path=f"screenshots/{endpoint}-{device_type}.png", full_page=True)
    browser.close()

with sync_playwright() as p:
    iphone_13 = p.devices['iPhone 13']
    browser = p.webkit.launch()
    context = browser.new_context(**iphone_13)
    take_screenshots("mobile", context)
    browser = p.webkit.launch()
    take_screenshots("desktop", browser)

File: src/ci/screenshots.ts
import fs from 'node:fs'
import { chromium, devices, type BrowserContext, type BrowserContextOptions } from 'playwright'

const baseUrl = process.argv[2]
const outDir = process.argv[3] || 'screenshots'
const endpoints = {
  Login: 'login',
  RouteList: '1d3dc3e03047b0c7',
  RouteActivity: '1d3dc3e03047b0c7/000000dd--455f14369d',
  SettingsActivity: '1d3dc3e03047b0c7/settings',
}

async function takeScreenshots(deviceType: string, context: BrowserContext) {
  const page = await context.newPage()
  for (const [route, path] of Object.entries(endpoints)) {
    await page.goto(`${baseUrl}/${path}`, { waitUntil: 'networkidle' })
    await page.waitForTimeout(1500)
    await page.screenshot({ path: `${outDir}/${route}-${deviceType}.playwright.png` })
    console.log(`${route}-${deviceType}.playwright.png`)

    if (route === 'Login') {
      await page.click("button:has-text('Try the demo')")
      await page.waitForLoadState('networkidle')
      await page.waitForTimeout(375)
    }
  }
  await page.close()
}

async function main() {
  let executablePath: string | undefined = '/usr/bin/chromium'
  if (!fs.existsSync(executablePath)) executablePath = undefined

  const browser = await chromium.launch({ executablePath, headless: true })

  const contexts: [string, BrowserContextOptions][] = [
    ['mobile', devices['iPhone 13']],
    ['desktop', { viewport: { width: 1920, height: 1080 } }],
  ]
  await Promise.all(
    contexts.map(async ([deviceType, options]) => {
      const context = await browser.newContext(options)
      await takeScreenshots(deviceType, context)
      await context.close()
    }),
  )

  await browser.close()
}

void main()
